pkg/gossip/gossip.go

```diff
@@ -1061,16 +1061,38 @@ func (g *Gossip) GetInfoStatus() InfoStatus {
 	return is
 }
 
-// IterateInfos visits all infos matching the given prefix.
-func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info) error) error {
+// CountInfos counts matching infos atomically under the lock. It is preferred
+// to using visitor for cases where actual info is irrelevant as it avoids
+// unnecessary cloning and holding a lock.
+func (g *Gossip) CountInfos(prefix string) int {
 	g.mu.RLock()
 	defer g.mu.RUnlock()
-	for k, v := range g.mu.is.Infos {
+	var count int
+	for k := range g.mu.is.Infos {
 		if strings.HasPrefix(k, prefix+separator) {
-			if err := visit(k, *(protoutil.Clone(v).(*Info))); err != nil {
-				return err
+			count++
+		}
+	}
+	return count
+}
+
+// IterateInfos visits all infos matching the given prefix. Function clones
+// matching records under lock, releases lock and then visits infos.
+func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info) error) error {
+	matching := make(map[string]Info)
+	func() {
+		g.mu.RLock()
+		defer g.mu.RUnlock()
+		for k, v := range g.mu.is.Infos {
+			if strings.HasPrefix(k, prefix+separator) {
+				matching[k] = *(protoutil.Clone(v).(*Info))
```

Comment:
That's fair, in that case it's tricky to move any of the logic outside the lock. I moved everything back as it was before and reverted the gossip change.
Current gossip interface is visitor based, but I don't think we need it in this form. We only ever use it for counting entries where we discard deserialized data or for queries where we retain the full set of data in some other form. None of those are particularly long running, but in general any long running op there breaks cluster because of liveness updates stalling.